teamcity-rest-client-impl/src/main/kotlin/org/jetbrains/teamcity/rest/coroutines/nodePinning.kt

```diff
@@ -0,0 +1,55 @@
+package org.jetbrains.teamcity.rest.coroutines
+
+import okhttp3.Cookie
+import okhttp3.CookieJar
+import okhttp3.HttpUrl
+import org.jetbrains.teamcity.rest.NodeSelector
+
+fun NodeSelector.toCookieJar(): CookieJar {
+    return when(this) {
+        is NodeSelector.PinToNode -> PinToNodeCookieJar(this.nodeId)
+        is NodeSelector.ServerControlled -> ServerControlledRouting()
+        is NodeSelector.Unspecified -> CookieJar.NO_COOKIES
+    }
+}
+
+private class ServerControlledRouting: CookieJar {
+    private var serverProvidedNodeId: String? = null
+    override fun loadForRequest(url: HttpUrl): List<Cookie> {
+        val targetNodeId = serverProvidedNodeId ?: return emptyList()
+
+        val domain = url.topPrivateDomain() ?: return emptyList()
+
+        return listOf(
+            Cookie.Builder()
+                .domain(domain)
+                .name("X-TeamCity-Node-Id-Cookie")
```

Comment:
I'd suggest extracting "X-TeamCity-Node-Id-Cookie" into a 
```kotlin
private const val TEAMCITY_NODE_ID_COOKIE = "X-TeamCity-Node-Id-Cookie"
```

To avoid duplication